### PR TITLE
Store initial blockchain state in TestAPIManager

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
This will enable dependency injection for #785.

### CODE REVIEW

File: source/agora/test/Base.d [hunk 1]
```diff
@@ -492,6 +498,34 @@ public interface TestAPI : ValidatorAPI
...
+
+    /// Load the blockchain
+    public override bool load ()
+    {
+        foreach (const ref block; this.blocks)
+            this.saveBlock(block);
+
+        return true;
+    }
+}
```
Comment (Geod24) [inline source/agora/test/Base.d]: Does it makes sense to have it here ? That could just replace the MemBlockStorage, couldn't it ? Do you remember why `MemBlockStorage` has serialized Blocks ?

Comment (AndrejMitrovic) [reply source/agora/test/Base.d]: > Does it makes sense to have it here ? 

Ah it should probably not be `public`. I can move it inside of `TestNodeMixin` I suppose. I only have one use-case for it and it's here in the tests.

> That could just replace the MemBlockStorage, couldn't it ? Do you remember why MemBlockStorage has serialized Blocks ?

Not sure. I think we can just remove that part? Then I could just add a constructor to `MemBlockStorage` and get rid of `PreLoadedMemBlockStorage` I suppose.

**Comment by AndrejMitrovic** [comment]
Ok I see why it's using serialization.

Because it needs to store `const(Block)` in `saveBlock`, but it needs to load mutable `Block` in e.g. `tryReadBlock`.

**Comment by AndrejMitrovic** [comment]
I've removed the extra class and just moved this stuff into the `MemBlockStorage` constructor.

File: source/agora/node/BlockStorage.d [hunk 2]
```diff
@@ -918,17 +918,23 @@ public class MemBlockStorage : IBlockStorage
...
+            foreach (const ref block; blocks)
+                this.saveBlock(block);
+        }
     }
 
     /// No-op: MemBlockStorage does no I/O
     public override bool load ()
     {
-        if (this.blocks.length == 0)
-            this.saveBlock(GenesisBlock);
```
Comment (Geod24) [inline source/agora/node/BlockStorage.d:931]: I would add a boolean in the new class, or at least something that checks we don't accidentally forget to initialize the BlockStorage from our code. [It happened before](https://github.com/bpfkorea/agora/commit/c7b90eb9483607d900fef63c3a7b63b2cae2b5dc)

**Comment by Geod24** [comment]
[This one too](https://github.com/bpfkorea/agora/pull/840/files#diff-b11d74d50f129e77728ec32bb7f9b4d9R1367) ?

File: source/agora/node/BlockStorage.d (hunk 2 above)
Comment (AndrejMitrovic) [reply source/agora/node/BlockStorage.d:931]: Ah then I'll move the loading into this method.

**Comment by AndrejMitrovic** [comment]
Split it into two commits, added a test-case for https://github.com/bpfkorea/agora/commit/c7b90eb9483607d900fef63c3a7b63b2cae2b5dc, and removed `PreloadedMemBlockStorage`.

File: source/agora/node/BlockStorage.d [hunk 3]
```diff
@@ -937,10 +949,6 @@ public class MemBlockStorage : IBlockStorage
         // Basic consistenty checks
         assert(this.height_idx.length == this.hash_idx.length);
         assert(this.height_idx.length == this.blocks.length);
-
-        // Make sure we have no empty block
-        foreach (blk; this.blocks)
-            assert(blk.length > 0);
```
Comment (Geod24) [inline source/agora/node/BlockStorage.d]: Is this test failing now ?

Comment (AndrejMitrovic) [reply source/agora/node/BlockStorage.d]: Ah I'll revert this, I initially replaced the serialized form with full blocks but that didn't work.

File: source/agora/test/Base.d [hunk 4]
```diff
@@ -543,6 +549,17 @@ private mixin template TestNodeMixin ()
...
+    private immutable(Block)[] blocks;
+
+    ///
+    public this (Config config, Registry* reg, immutable(Block)[] blocks)
+    {
+        this.registry = reg;
+        this.blocks = blocks;
+        super(config);
+    }
+
```
Comment (Geod24) [inline source/agora/test/Base.d:562]: Why aren't the ctor grouped together ?

Comment (AndrejMitrovic) [reply source/agora/test/Base.d:562]: Can you explain what you mean?

**Comment by Geod24** [comment]
Looks roughly good to me. Second commit title looks like a copy-pasta mistake though

**Comment by AndrejMitrovic** [comment]
Last commit message is wrong, I'll fix it.

File: source/agora/test/Base.d (hunk 4 above)
Comment (Geod24) [reply source/agora/test/Base.d:562]: I need some sleep

**Comment by AndrejMitrovic** [comment]
Fixed commits and reverted other change.

**Review by Geod24 — APPROVED** [review]

**Pull request merged**